File: src/examples/ex_gravity_comp.cpp
/* ex_gravity_comp.cpp
 * KDL gravity compensation on the Kinova GEN3: each step commands
 * KDL::ChainDynParam::JntToGravity in TORQUE mode, keeping the arm at its home pose.
 *
 * Usage:
 *   ex_gravity_comp [--headless]
 *
 * Runs 15 s (7500 steps) and prints the EE drift; --headless skips the viewer and exits 1 if
 * the drift exceeds kMaxDrift. */

#include "common.hpp"
#include "example_paths.hpp"

#include <iomanip>
#include <iostream>
#include <string>

namespace ex = mj_kdl_examples;

static constexpr int    kSteps    = 7500;   // 15 s
static constexpr double kMaxDrift = 0.0001; // [m]

int main(int argc, char *argv[])
{
    const bool headless = ex::parse_args(argc, argv).headless;

    mj_kdl::SceneSpec sc = ex::scene_spec();
    mj_kdl::RobotSpec r;
    r.path = ex::menagerie_model("kinova_gen3/gen3.xml");
    sc.robots.push_back(r);

    mj_kdl::Env   env;
    mj_kdl::Robot robot;
    if (!mj_kdl::init_env(&env, &sc)) return 1;
    if (!mj_kdl::init_robot_from_mjcf(&robot, &env, "base_link", "bracelet_link")) return 1;
    if (!mj_kdl::set_control_mode(&robot, mj_kdl::CtrlMode::TORQUE)) return 1;

    const unsigned                  n = robot.chain.getNrOfJoints();
    KDL::ChainFkSolverPos_recursive fk(robot.chain);
    KDL::ChainDynParam              dyn(robot.chain, KDL::Vector(0.0, 0.0, sc.gravity_z));
    const KDL::JntArray             q_home = ex::home_q(n);

    env.on_reset = [&](mj_kdl::ResetContext *) {
        mj_kdl::set_joint_pos(&robot, q_home);
        ex::prime_gravity(robot, dyn, q_home);
    };
    mj_kdl::reset(&env);
    if (!headless && !mj_kdl::open_viewer(&env)) return 1;

    KDL::Frame ee_start;
    fk.JntToCart(q_home, ee_start);
    for (int step = 0; step < kSteps; ++step) {
        mj_kdl::update(&env);
        ex::pd_gravity(robot, dyn, q_home);
        if (!mj_kdl::step(&env)) break;
        mj_kdl::pace_realtime(&env);
    }
    mj_kdl::update(&env);

    const double drift = (ex::tcp_frame(fk, robot).p - ee_start.p).Norm();
    std::cout << "EE drift after " << kSteps << " steps: " << std::fixed << std::setprecision(4)
              << drift * 1000.0 << " mm (limit " << kMaxDrift * 1000.0 << " mm)\n";
    mj_kdl::cleanup(&env);
    return headless ? ex::verdict(drift <= kMaxDrift, "the arm holds its pose") : 0;
}

File: python/mj_kdl_wrapper/examples/ex_gravity_comp.py
#!/usr/bin/env python3
"""KDL gravity compensation on the Gen3; Python counterpart of src/examples/ex_gravity_comp.cpp."""

from __future__ import annotations

import argparse

import PyKDL as kdl

import mj_kdl_wrapper as mjk

HOME_POSE = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]
STEPS = 7500  # 15 s
MAX_DRIFT = 0.0001  # [m]


def joints(values) -> kdl.JntArray:
    q = kdl.JntArray(len(values))
    for i, value in enumerate(values):
        q[i] = value
    return q


def gravity(dyn: kdl.ChainDynParam, q_values) -> list[float]:
    g = kdl.JntArray(len(q_values))
    dyn.JntToGravity(joints(q_values), g)
    return [g[i] for i in range(g.rows())]


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true")
    args = parser.parse_args()

    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True
    robot_spec = mjk.RobotSpec()
    robot_spec.path = mjk.menagerie.model_path("kinova_gen3", env_var="MJ_KDL_MODEL")
    spec.robots = [robot_spec]

    env = mjk.Env.build(spec)
    try:
        robot = env.create_robot("base_link", "bracelet_link")
        robot.set_control_mode(mjk.CtrlMode.TORQUE)
        chain = robot.kdl_chain()
        fk = kdl.ChainFkSolverPos_recursive(chain)
        dyn = kdl.ChainDynParam(chain, kdl.Vector(0.0, 0.0, spec.gravity_z))

        def on_reset(ctx):
            robot.set_joint_pos(HOME_POSE)
            robot.jnt_trq_cmd = gravity(dyn, HOME_POSE)

        env.on_reset = on_reset
        env.reset()
        if args.gui:
            env.open_viewer("ex_gravity_comp.py")

        start = kdl.Frame()
        fk.JntToCart(joints(HOME_POSE), start)
        for _ in range(STEPS):
            env.update()
            robot.jnt_trq_cmd = gravity(dyn, robot.jnt_pos_msr)
            if not env.step():
                break
            env.pace()
        env.update()

        end = kdl.Frame()
        fk.JntToCart(joints(robot.jnt_pos_msr), end)
        drift = (end.p - start.p).Norm()
        print(f"EE drift after {STEPS} steps: {drift * 1000:.4f} mm (limit {MAX_DRIFT * 1000} mm)")
    finally:
        env.close()
    if args.gui:
        return 0
    ok = drift <= MAX_DRIFT
    print(f"{'PASS' if ok else 'FAIL'}: the arm holds its pose")
    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main())
